src/vs/workbench/api/electron-browser/mainThreadStorage.ts

```diff
@@ -3,47 +3,74 @@
  *  Licensed under the MIT License. See License.txt in the project root for license information.
  *--------------------------------------------------------------------------------------------*/
 
-import { TPromise } from 'vs/base/common/winjs.base';
 import { IStorageService, StorageScope } from 'vs/platform/storage/common/storage';
-import { MainThreadStorageShape, MainContext, IExtHostContext } from '../node/extHost.protocol';
+import { MainThreadStorageShape, MainContext, IExtHostContext, ExtHostStorageShape, ExtHostContext } from '../node/extHost.protocol';
 import { extHostNamedCustomer } from 'vs/workbench/api/electron-browser/extHostCustomers';
+import { IDisposable } from 'vs/base/common/lifecycle';
 
 @extHostNamedCustomer(MainContext.MainThreadStorage)
 export class MainThreadStorage implements MainThreadStorageShape {
 
 	private _storageService: IStorageService;
+	private _proxy: ExtHostStorageShape;
+	private _storageListener: IDisposable;
+	private _sharedStorageKeysToWatch: Map<string, boolean> = new Map<string, boolean>();
 
 	constructor(
 		extHostContext: IExtHostContext,
 		@IStorageService storageService: IStorageService
 	) {
 		this._storageService = storageService;
+		this._proxy = extHostContext.getProxy(ExtHostContext.ExtHostStorage);
+
+		this._storageListener = this._storageService.onDidChangeStorage(e => {
+			let shared = e.scope === StorageScope.GLOBAL;
+			if (shared && this._sharedStorageKeysToWatch.has(e.key)) {
+				let value = this._getValue(shared, e.key);
+				if (value instanceof Error) {
+					return; // ignore parsing errors that can happen
+				}
+				this._proxy.$acceptValue(shared, e.key, value);
+			}
+		});
 	}
 
 	dispose(): void {
+		this._storageListener.dispose();
 	}
 
 	$getValue<T>(shared: boolean, key: string): Thenable<T> {
+		if (shared) {
+			this._sharedStorageKeysToWatch.set(key, true);
+		}
+		let value = this._getValue<T>(shared, key);
+		if (value instanceof Error) {
+			return Promise.reject(value);
+		}
+		return Promise.resolve(value);
+	}
+
+	private _getValue<T>(shared: boolean, key: string): T | Error {
 		let jsonValue = this._storageService.get(key, shared ? StorageScope.GLOBAL : StorageScope.WORKSPACE);
 		if (!jsonValue) {
-			return TPromise.as(undefined);
+			return undefined;
 		}
 		let value: T;
 		try {
 			value = JSON.parse(jsonValue);
-			return TPromise.as(value);
+			return value;
 		} catch (err) {
-			return TPromise.wrapError<T>(err);
+			return err;
```

Comment:
This is a very funny way of handling errors... I would do the try catch at the caller and do the resolve/reject there.

Reply:
@jrieken via https://github.com/Microsoft/vscode/pull/61024/commits/134ca026674e22afb967636443811d252c3e1076#diff-a5e120fd1c6388f23831d053bff4cdd6